src/main/groovy/com/github/jengelman/gradle/plugins/shadow/ShadowExtension.groovy

```diff
@@ -3,42 +3,50 @@ package com.github.jengelman.gradle.plugins.shadow
 import org.gradle.api.Project
 import org.gradle.api.artifacts.ProjectDependency
 import org.gradle.api.artifacts.SelfResolvingDependency
-import org.gradle.api.file.CopySpec
 import org.gradle.api.publish.maven.MavenPom
 import org.gradle.api.publish.maven.MavenPublication
 import org.gradle.util.GradleVersion
 
 class ShadowExtension {
 
-    CopySpec applicationDistribution
-    Project project
-
-    ShadowExtension(Project project) {
-        this.project = project
-        applicationDistribution = project.copySpec {}
-    }
-
-    void component(MavenPublication publication) {
+    void component(Project project, MavenPublication publication) {
 
         if (GradleVersion.current() >= GradleVersion.version("6.6")) {
             publication.artifact(project.tasks.named("shadowJar"))
         } else {
             publication.artifact(project.tasks.shadowJar)
         }
 
+        final def allDependencies = project.provider {
+            project.configurations.shadow.allDependencies.collect {
+                if ((it instanceof ProjectDependency) || ! (it instanceof SelfResolvingDependency)) {
+                    new Dep(it.group, it.name, it.version)
+                }
+            }
+        }
```

Comment:
Can we move `allDependencies` into `ShadowExtension`'s constructor?

Reply:
I would say yes if you don't put the `project` reference into a field.

Another trick you could use is to keep the `Project` field but as a `transient`.
It would then be available during the configuration phase but not anymore during execution.
This could allow to remove the need to have the `Project` parameter in the `component()` function.

Reply:
I tried it in https://github.com/Goooler/shadow/pull/17, but something is missing.

Reply:
It seems Groovy closure capturing is confused by the field instead of a local variable :/
It tries to find an xml element named `allDependencies` in the document :shrug: 
This looks like a Groovy problem to me.
Maybe declaring a local variable with the field assigned to it before the `withXml` closure would help :thinking:

Reply:
https://scans.gradle.com/s/iha6smzemxmuq/tests/:test/com.github.jengelman.gradle.plugins.shadow.PublishingSpec/publish%20shadow%20jar%20with%20maven-publish%20plugin?top-execution=1
or
https://scans.gradle.com/s/glxqqw5mj2rlm/tests/:test/com.github.jengelman.gradle.plugins.shadow.PublishingSpec/publish%20shadow%20jar%20with%20maven-publish%20plugin?top-execution=1

I was stuck there for days, I'm guessing there is something still not be serialized.

Reply:
Fixed by a86645c0dcb82fcd4663c313ed5345481478a85b.